### DISCUSSION TITLE
Train with known triplets

### DISCUSSION THREAD

**Discussion by swarmt** [KevinMusgrave/pytorch-metric-learning#695] [Q&A] [answered]
Hi all!

I feel like I am missing something obvious here, appreciate any guidance. 

I have a dataset that contains known image pairs. I want to train a model to minimise the cosine similarity between them.

How should I set up my data loader and training loop?

I think ideally I want to be able to specify the 3 images in the triplet or specify both the anchor positive and anchor negative pairs.

**Comment by KevinMusgrave** [comment, marked as answer]
Pairs or triplets are formed based on labels. So when you load a batch of image pairs, give each image in the pair the same label.

For example, let's say you have a batch of 4 images, where images 0 and 1 are a pair and images 2 and 3 are a pair:

```python
from pytorch_metric_learning.losses import ContrastiveLoss

loss_fn = ContrastiveLoss()

embeddings = model(images)
labels = torch.tensor([0,0,1,1])
loss = loss_fn(embeddings, labels)
```

If each image in each batch has exactly 1 positive, then you can accomplish the same thing with SelfSupervisedLoss without creating labels:

```python
from pytorch_metric_learning.losses import ContrastiveLoss, SelfSupervisedLoss

loss_fn = SelfSupervisedLoss(ContrastiveLoss())

anchor_embs = model(anchors)
positive_embs = model(positives)
loss = loss_fn(anchor_embs, positive_embs)
```

**Reply by swarmt** [reply]
That is excellent Kevin. Thank-you!

**Reply by swarmt** [reply]
This is what I came up with which appears to be working, I am getting a nice steady decrease in loss.

Is there a neat way to use the `AccuracyCalculator` with `SelfSupervisedLoss`?

```python

import numpy as np
import torchvision
from torch.utils.data import Subset

from particle.dataset import ImagePairDataset

import torch
import torch.nn as nn
import torch.optim as optim

from pytorch_metric_learning import testers
from pytorch_metric_learning.utils.accuracy_calculator import AccuracyCalculator
from pytorch_metric_learning.losses import ContrastiveLoss, SelfSupervisedLoss


class Net(nn.Module):
    def __init__(self, embedding_size=512, pretrained=True):
        super(Net, self).__init__()
        self.model = torchvision.models.resnet18(pretrained=pretrained)
        self.model.fc = nn.Linear(512, embedding_size)

    def forward(self, x):
        return nn.functional.normalize(self.model(x), dim=1)


def train(model, loss_func, device, train_loader, optimizer, epoch):
    model.train()
    for batch_idx, (anchor, positive) in enumerate(train_loader):
        anchor, positive = anchor.to(device), positive.to(device)
        optimizer.zero_grad()
        anchor_embs = model(anchor)
        positive_embs = model(positive)
        loss = loss_func(anchor_embs, positive_embs)
        loss.backward()
        optimizer.step()
        if batch_idx % 20 == 0:
            print(f"Epoch {epoch} Iteration {batch_idx}: Loss = {loss}")


def get_data_sets():
    samples = ImagePairDataset(input_size=224)
    n_samples = len(samples)
    n_train = int(0.8 * n_samples)
    train_indices = np.random.choice(n_samples, n_train, replace=False)
    val_indices = np.setdiff1d(np.arange(n_samples), train_indices)
    train_dataset = Subset(samples, train_indices)
    val_dataset = Subset(samples, val_indices)
    return train_dataset, val_dataset


if __name__ == "__main__":
    train_dataset, val_dataset = get_data_sets()
    train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=10, shuffle=True)
    val_loader = torch.utils.data.DataLoader(val_dataset, batch_size=1, shuffle=False)

    device = torch.device("cpu")
    loss_func = SelfSupervisedLoss(ContrastiveLoss())
    num_epochs = 10

    model = Net().to(device)
    optimizer = optim.Adam(model.parameters(), lr=0.0001)

    accuracy_calculator = AccuracyCalculator(include=("precision_at_1",), k=1)

    for epoch in range(1, num_epochs + 1):
        train(model, loss_func, device, train_loader, optimizer, epoch)
        # TODO: Implement validation

```

**Reply by KevinMusgrave** [reply]
There's no special integration of AccuracyCalculator with SelfSupervisedLoss. So you'll have to create labels that make sense "globally", i.e. every anchor-positive pair gets its own unique label within the validation set.

**Reply by swarmt** [reply]
I think I understand. Thanks for all the help Kevin and thanks for the fantastic library.

**Reply by swarmt** [reply]
Hey Kevin can I please check with you if this is correct?

My dataloader returns `anchor, positive, negative` tensors for each `__getitem__`

Where my mining_func is `EmbeddingsAlreadyPackagedAsTriplets()`

```python
    def compute_loss(self, batch):
        anchor, positive, negative = batch
        anchor_embeddings = self(anchor)
        positive_embeddings = self(positive)
        negative_embeddings = self(negative)
        embeddings = torch.cat([anchor_embeddings, positive_embeddings, negative_embeddings], dim=0)
        labels = torch.arange(embeddings.size(0))  # set labels to something, doesn't really matter
        triplets = self.mining_func(embeddings, labels)
        return self.loss_func(embeddings, labels, triplets)
```

**Reply by swarmt** [reply]
Or alternatively without the miner
```python
    def compute_loss(self, batch):
        anchor, positive, negative = batch
        combined = torch.cat([anchor, positive, negative], dim=0)
        embeddings = self(combined)
        batch_size = anchor.size(0)
        anchors = torch.arange(0, batch_size)
        positives = torch.arange(batch_size, 2 * batch_size)
        negatives = torch.arange(2 * batch_size, 3 * batch_size)
        indices_tuple = (anchors, positives, negatives)
        return self.loss_func(embeddings, labels=None, indices_tuple=indices_tuple)
```

**Reply by KevinMusgrave** [reply]
If I remember correctly, the [`EmbeddingsAlreadyPackagedAsTriplets`](https://kevinmusgrave.github.io/pytorch-metric-learning/miners/#embeddingsalreadypackagedastriplets) expects a slightly different format, so I would just create the `indices_tuple` yourself as you've done in your second snippet.

Your second snippet looks correct.
